internal/core/src/index/BitmapIndex.cpp

```diff
@@ -0,0 +1,648 @@
+                           "BitmapIndex can not build null values");
+    }
+
+    T* p = const_cast<T*>(data);
+    for (int i = 0; i < n; ++i, ++p) {
+        data_[*p].add(i);
+    }
+    total_num_rows_ = n;
+
+    for (auto it = data_.begin(); it != data_.end(); ++it) {
+        bitsets_[it->first] = ConvertRoaringToBitset(it->second);
+    }
+
+    is_built_ = true;
+}
+
+template <typename T>
+void
+BitmapIndex<T>::BuildV2(const Config& config) {
+    if (is_built_) {
+        return;
+    }
+    auto field_name = file_manager_->GetIndexMeta().field_name;
+    auto reader = space_->ScanData();
+    std::vector<FieldDataPtr> field_datas;
+    for (auto rec = reader->Next(); rec != nullptr; rec = reader->Next()) {
+        if (!rec.ok()) {
+            PanicInfo(DataFormatBroken, "failed to read data");
+        }
+        auto data = rec.ValueUnsafe();
+        auto total_num_rows = data->num_rows();
+        auto col_data = data->GetColumnByName(field_name);
+        auto field_data = storage::CreateFieldData(
+            DataType(GetDType<T>()), 0, total_num_rows);
+        field_data->FillFieldData(col_data);
+        field_datas.push_back(field_data);
+    }
+
+    int total_num_rows = 0;
+    for (auto& field_data : field_datas) {
+        total_num_rows += field_data->get_num_rows();
+    }
+    if (total_num_rows == 0) {
+        throw SegcoreError(DataIsEmpty,
+                           "scalar bitmap index can not build null values");
+    }
+
+    total_num_rows_ = total_num_rows;
+
+    int64_t offset = 0;
+    for (const auto& data : field_datas) {
+        auto slice_row_num = data->get_num_rows();
+        for (size_t i = 0; i < slice_row_num; ++i) {
+            auto val = reinterpret_cast<const T*>(data->RawValue(i));
+            data_[*val].add(offset);
+            offset++;
+        }
+    }
+    is_built_ = true;
+}
+
+template <typename T>
+void
+BitmapIndex<T>::SerializeIndexData(uint8_t* data_ptr) {
+    for (auto& pair : data_) {
+        memcpy(data_ptr, &pair.first, sizeof(T));
+        data_ptr += sizeof(T);
+
+        pair.second.write(reinterpret_cast<char*>(data_ptr));
+        data_ptr += pair.second.getSizeInBytes();
+    }
+}
+
+template <>
+void
+BitmapIndex<std::string>::SerializeIndexData(uint8_t* data_ptr) {
+    for (auto& pair : data_) {
+        size_t key_size = pair.first.size();
+        memcpy(data_ptr, &key_size, sizeof(size_t));
+        data_ptr += sizeof(size_t);
+
+        memcpy(data_ptr, pair.first.data(), key_size);
+        data_ptr += key_size;
+
+        pair.second.write(reinterpret_cast<char*>(data_ptr));
+        data_ptr += pair.second.getSizeInBytes();
+    }
+}
+
+template <typename T>
+BinarySet
+BitmapIndex<T>::Serialize(const Config& config) {
+    AssertInfo(is_built_, "index has not been built yet");
+
+    auto index_data_size = 0;
+    for (auto& pair : data_) {
+        index_data_size += pair.second.getSizeInBytes() + sizeof(T);
```

Comment:
I think sizeof(T) is not correct when T is std::string, right?

Reply:
good catch, it is error for judge size of index data, I will using template specialization to refactor for string